server/app/services/applicant/ApplicantService.java

```diff
@@ -665,6 +674,22 @@ private void logDuplicateDrafts(ImmutableSet<Application> applications) {
     }
   }
 
+  private int getAutoFillCount(ReadOnlyApplicantProgramService roApplicantService) {
+    return roApplicantService.getSummaryData().stream().filter(data -> data.isPreviousResponse()).collect(ImmutableList.toImmutableList()).size();
+  }
+
+  private int getAutoFillCount(Application application) {
+    ReadOnlyApplicantProgramService roApplicantService = getReadOnlyApplicantProgramService(application).toCompletableFuture().join();
```

Comment:
suggestion (blocking): avoid calling `join` here because it blocks the HTTP thread on database IO, which we want to avoid doing for any applicant-facing routes.

Two possible approaches for improving this:
1. do the IO here asynchronously and make this method and its caller asynchronous (change their return type to `CompletionStage<T>`)
1. collect the data needed for this method to do its work somewhere else asynchronously so that this code accesses local data only and stays synchronous

Reply:
@bion ,
Got back to this and I ended going with the first approach, since it introduced the least amount of complexity. I'm sure we can call `join` less than what I have here, but that requires more overhaul, so let me know what you think.

Reply:
> I'm sure we can call join less than what I have here, but that requires more overhaul, so let me know what you think.

Unfortunately that's what's required to maintain [the standard of not blocking HTTP threads in applicant-facing routes](https://docs.civiform.us/contributor-guide/developer-guide/development-standards#async-request-handling).